machineconfiguration/v1alpha1/types_machineosbuild.go

```diff
@@ -0,0 +1,131 @@
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata"`
+
+	Items []MachineOSBuild `json:"items"`
+}
+
+// MachineOSBuildSpec describes user-configurable options as well as information about a build process.
+type MachineOSBuildSpec struct {
+	// machineConfigPool is the pool which the build is for
+	// +kubebuilder:validation:Required
+	MachineConfigPool MachineOSObjectReference `json:"machineConfigPool"`
+	// renderedMachineOSImage describes the machineOsImage object created to track image specific information
+	// machineConfig is a reference to the MC that was used to build this image.
+	// +optional
+	RenderedMachineOSImage MachineOSObjectReference `json:"renderedMachineOSImage"`
+	// buildInputs is where user options for the build live
+	// +kubebuilder:validation:Required
+	BuildInputs BuildInputs `json:"buildInputs"`
+	// currentConfig is the currently running config on the MCP
+	// +optional
+	CurrentConfig string `json:"currentConfig"`
+	// desiredConfig is the desired config we want to build an image for. If currentConfig and desiredConfig are not the same, we need to build an image.
+	// +kubebuilder:validation:Required
+	DesiredConfig string `json:"desiredConfig"`
+}
+
+// MachineOSBuildStatus describes the state of a build and other helpful information.
+type MachineOSBuildStatus struct {
+	// buildStart describes when this build began
+	// +kubebuilder:validation:Required
+	BuildStartTime metav1.Time `json:"buildStart"`
+	// buildEnd describes when the build ended
+	// +optional
+	BuildEndTime metav1.Time `json:"buildEnd"`
+	// conditions are state related conditions for the build. Valid types are:
+	// BuildPrepared, Building, BuildFailed, BuildInterrupted, BuildRestarted, and Built
+	// +patchMergeKey=type
+	// +patchStrategy=merge
+	// +listType=map
+	// +listMapKey=type
+	// +optional
+	Conditions []metav1.Condition `json:"conditions,omitempty" patchStrategy:"merge" patchMergeKey:"type"`
+	// relatedBuilds contains previous iterations of failed or interrupted builds related to this one.
+	// This will keep track of failed or interrupted build names
+	// I am thinking we keep track of their name and why they failed
+	// +listType=set
+	// +optional
+	RelatedBuilds []PriorMachineOSBuilds `json:"relatedBuilds,omitempty"`
+}
+
+// BuildProgess highlights some of the key phases of a build to be tracked in Conditions.
+type BuildProgress string
+
+const (
+	MachineOSBuildPrepared    BuildProgress = "BuildPrepared"
+	MachineOSBuiling          BuildProgress = "Building"
+	MachineOSBuildFailed      BuildProgress = "BuildFailed"
+	MachineOSBuildInterrupted BuildProgress = "BuildInterrupted"
+	MachineOSBuildRestarted   BuildProgress = "BuildRestarted"
+	MachineOSBuilt            BuildProgress = "Built"
+)
+
+// PriorMachineOSBuilds contains information about related builds
+type PriorMachineOSBuilds struct {
+	// name is the name of the build
+	// +kubebuilder:validation:Required
+	Name string `json:"name"`
```

Comment:
Oh good question, I had assumed from where this is in spec, that it was another object, but if it's pods then tracking that is definitely like a Job

Reply:
this tracks old MachineOSBuilds (they will still be around until they are rotten)

Reply:
Just to make sure I understand correctly:

We have a MachineOSBuild object. Whenever something changes, we're creating a child MachineOSBuild object? Kind of like what we're doing with MachineConfigs and rendered MachineConfigs? Or am I misunderstanding this?

I was under the impression that a MachineOSBuild object holds the needed configuration(s) for a pool to do a build. Whenever certain elements of the configuration change, the MachineOSBuilder controller detects that change and then launches a build. This would then produce a MachineOSImage (which I know is beyond the scope of this PR). And whenever a new MachineOSImage is produced, that gets assigned to a MachineConfigPool so that we can roll it out.

Reply:
Not a fan of child objects (sorry MachineConfig pattern), but yes, I was on Zack's understanding as well

Reply:
Yeah, I think that's the best way forward.